CODING_GUIDELINES.md

```diff
@@ -0,0 +1,690 @@
+This means that it is important to verify that the entries are actually identical, and just not that they hash to the same value.
+
+This is especially important when hashing attacker-controlled values, as they can be specially crafted to trigger collisions to cause:
+
+* cache pollution (see [2018-06: Packet cache pollution via crafted query](https://docs.powerdns.com/recursor/security-advisories/powerdns-advisory-2018-06.html)) ;
+* denial of service via hash table flooding (in a map, all entries that hash to the same value are often placed into a linked-list, making it possible to cause a linear scan of entries by making all of them hash to that same value).
+
+The first issue can be prevented by comparing the entries and not just the value they hash to.
+The second one can be used by using some sort of secret when computing the hash so that the result cannot be guessed by the attacker.
+That can be achieved by using an unpredictable seed for certain hash algorithms, or a secret for some other like `SipHash`.
+
+# Readability tips
+
+Some of these tips are actually enforced by `clang-tidy` nowadays, but it is still useful to keep them in mind.
+
+## auto
+
+C++11 introduced automatic type deduction, using the `auto` keyword.
+In addition to saving the typing of a few more letters, using automatic type deduction prevents nasty surprises if the variable is initialized from another one, or from a function, and the other type is changed to a different one.
+The code might still compile while now involving a copy or worse.
+
+## Boolean expressions
+
+## Explicit comparisons
+
+* compare numerical values with `== 0` or `!= 0` explicitly ;
+* compare to `false` explicitly, which is easier to read ;
+* compare to `nullptr` for the same reason.
+
+## Initialization
+
+Use braced initialization for members as often as possible:
+
+* it does forbid narrowing conversions
+* and avoids C++'s "[most vexing parse](https://en.wikipedia.org/wiki/Most_vexing_parse)" which is to declare a function instead of calling the default constructor:
+
+```c++
+Object a(); // declares a function named a that returns an object
+```
+
+## nullptr
+
+When representing a pointer, using `nullptr` makes it immediately obvious that we are dealing with a pointer, as opposed to the use of `0`.
+It also cannot be silently taken as an integer, which can happens with `0` but also with `NULL`.
+
+## const-ness
+
+* Mark parameters and variables that should not be modified as `const`.
+  This is especially important for references and pointers that comes from outside the function, but it also makes sense to do it for local variables or parameters passed by value because it might help detect a logic error later.
+* Mark const methods as such (and make them thread-safe)
+* Prefer using `at()` on containers so that no insertion can take place by mistake, and to get bounds checking.
+
+## static
+
+Functions that are only used inside a single file should be marked as `static`, so that:
```

Comment:
I think we prefer `namespace { ... }` these days?

Reply:
I'll mention both options since I don't really have an opinion on whether one is better than the other, but perhaps you or Fred do?

Reply:
I have a small preference for `namespace { ... }`

Reply:
Same here.

Reply:
Alright!